File: crack.c
#include <stdio.h>
#include <cs50.h>
#include <string.h>
#define _XOPEN_SOURCE
#include <unistd.h>
#include <crypt.h>

int main(int argc, string argv[])
{
    if (argc != 2)
    {
        printf("Usage: crack <hashed password>\n");
        return 1;
    }

    char salt[2];
    salt[0] = argv[1][0];
    salt[1] = argv[1][1];

    char testKey[5];
    testKey[1] = '\0';

    //check single-character passwords
    for (int i = 65; i <= 122; i++)
    {
        //skip the characters between upper and lowercase letters
        if (i == 91)
        {
            i = 97;
        }

        testKey[0] = i;

        if (strcmp(crypt(testKey, salt), argv[1]) == 0)
        {
            printf("%s\n", testKey);
            return 0;
        }
    }

    //check two-character passwords
    testKey[2] = '\0';
    for (int i = 65; i <= 122; i++)
    {
        //skip the characters between upper and lowercase letters
        if (i == 91)
        {
            i = 97;
        }

        testKey[0] = i;
        for (int j = 65; j <= 122; j++)
        {
            //skip the characters between upper and lowercase letters
            if (j == 91)
            {
                j = 97;
            }

            testKey[1] = j;

            if (strcmp(crypt(testKey, salt), argv[1]) == 0)
            {
                printf("%s\n", testKey);
                return 0;
            }
        }
    }

    //check three-character passwords
    testKey[3] = '\0';
    for (int i = 65; i <= 122; i++)
    {
        //skip the characters between upper and lowercase letters
        if (i == 91)
        {
            i = 97;
        }

        testKey[0] = i;
        for (int j = 65; j <= 122; j++)
        {
            //skip the characters between upper and lowercase letters
            if (j == 91)
            {
                j = 97;
            }

            testKey[1] = j;
            for (int k = 65; k <= 122; k++)
            {
                //skip the characters between upper and lowercase letters
                if (k == 91)
                {
                    k = 97;
                }

                testKey[2] = k;

                if (strcmp(crypt(testKey, salt), argv[1]) == 0)
                {
                    printf("%s\n", testKey);
                    return 0;
                }
            }
        }
    }

    //check four-character passwords
    testKey[4] = '\0';
    for (int i = 65; i <= 122; i++)
    {
        //skip the characters between upper and lowercase letters
        if (i == 91)
        {
            i = 97;
        }

        testKey[0] = i;
        for (int j = 65; j <= 122; j++)
        {
            //skip the characters between upper and lowercase letters
            if (j == 91)
            {
                j = 97;
            }

            testKey[1] = j;
            for (int k = 65; k <= 122; k++)
            {
                //skip the characters between upper and lowercase letters
                if (k == 91)
                {
                    k = 97;
                }

                testKey[2] = k;
                for (int l = 65; l <= 122; l++)
                {
                    //skip the characters between upper and lowercase letters
                    if (l == 91)
                    {
                        l = 97;
                    }

                    testKey[3] = l;

                    if (strcmp(crypt(testKey, salt), argv[1]) == 0)
                    {
                        printf("%s\n", testKey);
                        return 0;
                    }
                }
            }
        }
    }
}

File: crack.py
# This implementation is less efficient than it could be, given how constrained the specifications are. However, it is significantly
# more generalized: by editing the two global constants, a much wider variety of passwords can be cracked.

import sys
import crypt

VALID_PASSWORD_CHARACTERS = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ"
MAX_PASSWORD_LENGTH = 4


def main():
    # check for correct number of command line arguments
    if len(sys.argv) != 2:
        print("usage: crack hashed_password")
        exit(1)

    target = sys.argv[1]
    salt = target[0:2] # crypt stores the salt as the first two characters in the hash

    for i in range(MAX_PASSWORD_LENGTH):
        passwords = generate_possible_passwords(VALID_PASSWORD_CHARACTERS, i)
        for password in passwords:
            if crypt.crypt(password, salt) == target:
                print(password)
                exit(0)

    exit(2)


# Recursize function that generates a list of all possible passwords of a given length. This has the potential to use an extrmeely
# high amount of memory when given a long length, but for the short passwords called for in the specifications, it is fine. This
# approach is used in the interest of making the code more readable and understandable.
def generate_possible_passwords(passwords, length):
    if length == 0:
        return passwords
    else:
        newPasswords = []
        for password in passwords:
            for char in VALID_PASSWORD_CHARACTERS:
                newPasswords.append(str(password + char))
        return generate_possible_passwords(newPasswords, length - 1)


if __name__ == "__main__":
    main()